Issue #734: Context menu & menu bar edits

katelynienaber opened the pull request:
## Proposed changes

Small changes to context menus & menu bars according to conversation with @jelaplan and the team:
- Moved all commands into the context menu except: Add Session, Refresh All, Search, Add/Remove Favorite
- Renamed some commands to be more specific: Paste/Paste Data Set, Issue Command/Issue MVS Command (as requested by Venkat & John)
- Put Search at the top of every menu
- Put Update/Hide/Delete Profile at the bottom of every menu
- Gave Add to Favorites its own group in the menu
- Reduced number of groups in the menus...we now have DS/file options, auth options, & profile options (plus some small groups specific to certain menus)

@lauren-li I tagged you for review because you had helped with reordering the context menus before :)

## Release Notes
Milestone: Backlog
Changelog: All commands moved/copied into the context menus, and some commands renamed to be more specific.

## Types of changes

What types of changes does your code introduce to Zowe Explorer?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Updates to Documentation or Tests (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This checklist will be used as reference for both the contributor and the reviewer_

- [x] I have read the [CONTRIBUTOR GUIDANCE](https://github.com/zowe/vscode-extension-for-zowe/wiki/Best-Practices:-Contributor-Guidance) wiki
- [x] PR title follows [Conventional Commits Guidelines](https://www.conventionalcommits.org/en/v1.0.0-beta.2/)
- [x] PR Description is included
- [x] `npm run vscode:prepublish` has been executed
- [x] All checks have passed (DCO, Jenkins and Code Coverage)
- [ ] I have added unit test and it is passing
- [ ] I have added integration test and it is passing
- [ ] There is coverage for the code that I have added
- [x] I have tested it manually and there are no regressions found
- [x] I have added necessary documentation (if appropriate)
- [x] Any PR dependencies have been merged and published (if appropriate)

## Further comments

This PR does NOT close Issue #734. It just happens to be related. **So please don't close the issue!** :D

jellypuno reviewed (approved):
Looks Great! Just one comment on the Edit Label. Since all functions has the word dataset or member in it, maybe you can add Dataset and Member in the Edit as well. Just for consistency. Same with Rename and Delete File. Other than that, everything looks good!

![image](https://user-images.githubusercontent.com/45042840/91832324-8ebef800-ec45-11ea-9809-5670bb7834f5.png)

katelynienaber commented:
> Looks Great! Just one comment on the Edit Label. Since all functions has the word dataset or member in it, maybe you can add Dataset and Member in the Edit as well. Just for consistency. Same with Rename and Delete File. Other than that, everything looks good!
> 
> ![image](https://user-images.githubusercontent.com/45042840/91832324-8ebef800-ec45-11ea-9809-5670bb7834f5.png)

Fixed this for Copy & Paste Data Set/Member

jellypuno commented:
Was this handled?

![image](https://user-images.githubusercontent.com/45042840/91840620-4574a580-ec51-11ea-8220-eb754c741124.png)

katelynienaber commented:
> Was this handled?
> 

I just added it in the latest version :)

jellypuno reviewed (approved):
Looking good! I just have this error. Maybe it is from my machine. Please check if this error is not happening to others.

![image](https://user-images.githubusercontent.com/45042840/92736899-f5958d00-f37a-11ea-8b36-a9d8df4a17b4.png)

lauren-li reviewed (dismissed):
Thanks @katelynienaber for the updates in this PR!

Apologies for being slow on this review. It took me some time to get what was happening with the changes to DS createFile and USS createUSSNode and why, but I understand now.

Overall, the PR looks good! All unit and expected integration tests are passing. I think it is near ready for approval, but I have some quick suggestions and questions that I think should be addressed before this goes into master branch (see code references).

I know you are focusing heavily on the base profile PR now, so I tried to figure out some of the issues to make polishing this PR quicker and easier for you. Hope it helps!

lauren-li commented:
@zFernand0 @katelynienaber Just a heads-up when bringing this branch up-to-date with the master branch: The context menu commands for favoriting a USS session and favoriting a Jobs session are now `zowe.uss.addFavorite` and `zowe.jobs.addFavorite`, respectively (instead of `x.saveSearch`). (The functionality to favorite/save a search was moved into addFavorite since they shared so much code.)

Therefore, you may encounter a merge conflict in package.json that looks like:
<img width="769" alt="Screen Shot 2020-09-11 at 12 21 15 PM" src="https://user-images.githubusercontent.com/45975633/93009650-8756f300-f551-11ea-8ee4-af3ae68923c7.png">

It should be resolved to look something like below. (Note especially the `command` value: `zowe.uss.addFavorite`.):
```
       {
          "when": "view == zowe.uss.explorer && viewItem =~ /^(?!.*_fav.*)ussSession.*/",
          "command": "zowe.uss.addFavorite",
          "group": "3_workspace@3"
        },
```

katelynienaber commented on package.json:
```diff
@@ -543,48 +543,23 @@
         },
         {
-          "when": "view == zowe.uss.explorer && viewItem =~ /^directory.*/",
+          "when": "view == zowe.uss.explorer && viewItem =~ /^directory|ussSession/",
```
Fixed this, maybe we can add back sessions in faves for this command later

katelynienaber commented on package.json:
```diff
@@ -1351,7 +1338,7 @@
   },
   "dependencies": {
-    "@zowe/cli": "^6.20.0",
+    "@zowe/cli": "^6.22.0",
```
I reverted this, I don't think it was necessary

katelynienaber commented:
@jellypuno @lauren-li I think I have gotten to your comments...it should be reviewed again ^^'. But it's sort of whenever because @jellypuno is on vacation and anyway other stuff is going into this release so there's ages to wait on it.

jellypuno commented on .github/workflows/theia-zowe-explorer-ci.yml:
```diff
@@ -12,14 +12,22 @@ jobs:
+      # Don't think this is necessary, GH Actions docs says
+      # yarn is installed by default:
+      # https://docs.github.com/en/free-pro-team@latest/actions/guides/building-and-testing-nodejs#installing-dependencies
+      # - run: npm install -g yarn
```
do not remove this. Just add sudo:  `sudo npm install -g yarn`

katelynienaber replied:
@jellypuno I will remove it, but could you explain why? Are we not using GH-hosted runners? Those have yarn installed by default apparently.

jelaplan commented:
I thikn that copy and paste refer to members. You can pick a member and choose copy. Then you click on a data set and click paste which then pastes the copied member into that data set. We may get copy and paste data sets in the future but for now we can not. We can allocate like a data set or create a new data set. 
![image](https://user-images.githubusercontent.com/3207260/98738488-c64ccd80-2375-11eb-9f24-e8f9a31484f3.png)

katelynienaber commented:
> I think that copy and paste refer to members. You can pick a member and choose copy. Then you click on a data set and click paste which then pastes the copied member into that data set. We may get copy and paste data sets in the future but for now we can not. We can allocate like a data set or create a new data set.
> ![image](https://user-images.githubusercontent.com/3207260/98738488-c64ccd80-2375-11eb-9f24-e8f9a31484f3.png)

@jelaplan So, I checked this in the function, and apparently "Copy Data Set/Copy Member" actually refers to the data set/member name. Therefore I did the following:
- When the user right-clicks on a Data Set, they will see the option "Copy Data Set Name"
- When the user right-clicks on a Member, they will see "Copy Member" AND new option "Copy Member Name"

If users want to copy the member contents, they can just open the member. They might still be confused, if they click Copy Member and try to paste directly, and they get some weird JSON object with just DS name and profile name. I am not sure how to fix this...maybe @jelaplan has a better idea for UX here?

**For other reviewers**, I have changed the name of function _pasteDataSet_ to _pasteMember_, since you can't paste a data set anyway.

**Also**, new function Copy Member Name now actually returns the member name, instead of some weird JSON object containing parent data set name and profile name :)

lauren-li commented:
@katelynienaber Just sticking a request here for when we want to add this in (for v2.0?): For consistency with VS Code's context menus, we should try to keep Hide/Rename/Delete at the bottom of Zowe Explorer's context menus. For files/folders (or data sets/members), I think we have already implemented this, but for profiles, we are starting to add quite a few items underneath Delete:
<img width="277" alt="Screen Shot 2020-12-03 at 7 16 23 AM" src="https://user-images.githubusercontent.com/45975633/101017766-87570580-3538-11eb-881f-417533983015.png">

If we agree on trying to keep Hide/Rename/Delete at botttom of the context menus, perhaps we should use high numbers like `modification@97`, `modification@98`, `modification@99` for these options in `package.json` so developers don't have to keep remembering to modify the ordering whenever a new context menu option is added into the `modification` category.

katelynienaber commented:
> @katelynienaber Just sticking a request here for when we want to add this in (for v2.0?): For consistency with VS Code's context menus, we should try to keep Hide/Rename/Delete at the bottom of Zowe Explorer's context menus. For files/folders (or data sets/members), I think we have already implemented this, but for profiles, we are starting to add quite a few items underneath Delete:
> <img alt="Screen Shot 2020-12-03 at 7 16 23 AM" width="277" src="https://user-images.githubusercontent.com/45975633/101017766-87570580-3538-11eb-881f-417533983015.png">
> 
> If we agree on trying to keep Hide/Rename/Delete at botttom of the context menus, perhaps we should use high numbers like `modification@97`, `modification@98`, `modification@99` for these options in `package.json` so developers don't have to keep remembering to modify the ordering whenever a new context menu option is added into the `modification` category.

Last commit addresses this comment.

lauren-li commented on package-lock.json:
```diff
@@ -0,0 +1,12963 @@
+{
```
Since our repo now uses Yarn, we should probably remove this `package-lock.json` file.

lauren-li commented on yarn.lock:
```diff
@@ -11579,4 +11579,4 @@ zos-node-accessor@1.0.7:
     debug "~2.6.8"
     ftp4 "~0.3.13"
-    q "~1.4.1"
+    q "~1.4.1"
```
Can this change be removed from the PR? This PR doesn't seem like it should involve any changes to dependencies or the `lock` file.

katelynienaber replied:
This one is from Master, if I remove it the tests break P:

lauren-li commented on src/uss/USSTree.ts:
```diff
@@ -0,0 +1,529 @@
+/*
```
This file is probably a leftover from when you merged in the monorepo changes, and should be removed.

lauren-li commented:
@katelynienaber I like the changes I see in the UI when manually testing so far, but it seems like some changed files/folders in this PR need to be moved into their respective locations in `packages/zowe-explorer` (or removed from the PR if there are no changes from the `master` branch). For example:
- `__mocks__/**`
- `i18n/**`
- `package.nls.json`

katelynienaber commented:
@lauren-li I made these changes, please take another look ^^'

jellypuno commented:
@katelynienaber I was testing this and I noticed that after I click search it will issue lots of stuff as well. There's even an error about issue command which doesn't happen in master.

![ContextMenu](https://user-images.githubusercontent.com/45042840/104616267-5f57d780-568a-11eb-9888-78b07daf07c3.gif)

katelynienaber commented:
> @katelynienaber I was testing this and I noticed that after I click search it will issue lots of stuff as well. There's even an error about issue command which doesn't happen in master.
> 
> ![ContextMenu](https://user-images.githubusercontent.com/45042840/104616267-5f57d780-568a-11eb-9888-78b07daf07c3.gif)

@jellypuno I think I messaged you about this...anyway the only thing I changed in that function was some text in a user prompt. The correct function is being called from the context menu, and I haven't changed the logic in there at all.

If there is an issue with how data sets are searched I think it should be addressed as another issue. Because the multi-loading does occur in Master. I am not sure about that error but it's a MF thing. Maybe someone else can check on this as well.

jellypuno commented:
>  I am not sure about that error but it's a MF thing, not some VSCode error. Maybe someone else can check on this as well.

This is not an MF thing. The API calls are triggered for some reason after loading the list of datasets. I'll debug it.

lauren-li commented:
> This is not an MF thing. The API calls are triggered for some reason after loading the list of datasets. I'll debug it.

We discussed this in the Slack channel, but just for the record of this conversation: It seems the issue started with commit e4456e7 in PR #1072 (PDS Member Search filter). The `refreshElement` line was added to fix an issue I had mentioned in the [preceding comment](https://github.com/zowe/vscode-extension-for-zowe/pull/1072#pullrequestreview-533838851). (This is in `DatasetTree.ts`):

<img width="885" alt="Screen Shot 2021-01-18 at 10 07 43 AM" src="https://user-images.githubusercontent.com/45975633/105027920-5bc3a800-5a1e-11eb-8e7b-2e03affa827a.png">

So, we need to implement a way to refresh the PDS content when the user does another member search on it- without calling refreshElement on all the data sets returned. We might also want to consider moving this logic or adding a condition so that it only happens for member filtering searches, and all general data set searches too.

This should probably be addressed in a separate PR, but good catch @jellypuno! I'm glad this issue was found before we released it.

lauren-li commented on packages/zowe-explorer/src/uss/actions.ts:
```diff
@@ -47,18 +50,32 @@ export async function createUSSNode(
     nodeType: string,
     isTopLevel?: boolean
 ) {
+    await ussFileProvider.checkCurrentProfile(node);
```
The changes in this file deal with creating a new USS file or folder, which seems unrelated to the rest of this PR. It does seem to work well, so I guess we can leave it in this time, but going forward, I would recommend making a separate PR for separate fixes/features.

This adds an extra step in the folder/file creation process, but it does give the user a more obvious way to control where they are creating the file/folder, and helps avoid errors if the user wants to do the creation without having entered a search prompt first. The fact that it's pre-filled makes helps it more painless. I like that it gets rid of the USS tree collapsing upon node creation.

katelynienaber replied:
This was added to fix a bug with using the commands via the context menu, instead of via the button on the toolbar...it should also allow us to create new through the command palette later on, but that isn't in this PR.

lauren-li reviewed (changes requested):
Overall, this PR seems in good shape. All unit and expected integration tests are passing. However, the favorites star icon is gone in the Data Sets view (for sequential and partioned data sets, as well as data set members), while it remains for USS and Jobs. Whether we decide to keep the icon or move the command exclusively to the context menu, it should be consistent across the 3 different views:
![ze-favstar4](https://user-images.githubusercontent.com/45975633/105043086-08a82000-5a33-11eb-87af-973faa60ed61.gif)

Additionally, the option to add a profile/search to Favorites in the Data Sets view appears to have been removed from the context menu, so there is no longer any way to add these to Data Set Favorites:
<img width="350" alt="Screen Shot 2021-01-19 at 8 21 24 AM" src="https://user-images.githubusercontent.com/45975633/105042913-d39bcd80-5a32-11eb-88d8-fabc07880277.png">

Aside from that, the other commands that had icons before appear to all have been successfully moved into their respective context menus. Some points of consideration:
- I actually found the trash can icon for Jobs to be quite useful, but maybe it's okay to remove the icon, especially if we are working towards a multi-select feature so users can delete multiple jobs in one go.
- @katelynienaber It might be useful for us to put together a wiki page for developers regarding what the context menu category options and ordering for commands are. I think this would help our developers keep the context menus organized when adding new commands, as opposed to just adding them all to the bottom of the menu (which is what tends to happen currently).

katelynienaber commented:
> Overall, this PR seems in good shape. All unit and expected integration tests are passing. However, the favorites star icon is gone in the Data Sets view (for sequential and partioned data sets, as well as data set members), while it remains for USS and Jobs. Whether we decide to keep the icon or move the command exclusively to the context menu, it should be consistent across the 3 different views:

I put this back.

> Additionally, the option to add a profile/search to Favorites in the Data Sets view appears to have been removed from the context menu, so there is no longer any way to add these to Data Set Favorites:

Also returned.

> Aside from that, the other commands that had icons before appear to all have been successfully moved into their respective context menus. Some points of consideration:
> 
> * I actually found the trash can icon for Jobs to be quite useful, but maybe it's okay to remove the icon, especially if we are working towards a multi-select feature so users can delete multiple jobs in one go.

Can we maybe wait till the multi-delete is implemented and see if it's ok to leave the icon off? I already agreed with John about removing it XD

> * @katelynienaber It might be useful for us to put together a wiki page for developers regarding what the context menu category options and ordering for commands are. I think this would help our developers keep the context menus organized when adding new commands, as opposed to just adding them all to the bottom of the menu (which is what tends to happen currently).

@zdmullen What do you think about this last point? Where should I set it up?

jellypuno commented:
> @zdmullen What do you think about this last point? Where should I set it up?

We can create a wiki here in our repo or a blog? 😉 Better..... video blog 😄 kidding... unless....

lauren-li commented:
> >  I actually found the trash can icon for Jobs to be quite useful, but maybe it's okay to remove the icon, especially if we are working towards a multi-select feature so users can delete multiple jobs in one go.
>
>Can we maybe wait till the multi-delete is implemented and see if it's ok to leave the icon off? I already agreed with John about removing it XD

@katelynienaber I am ok with removing the trash can icon in this PR, as I believe the squad already agreed to removing it in earlier discussions. Once multi-select is in for deleting jobs, it should more than make up for the missing trash cans.

~
>  > @katelynienaber It might be useful for us to put together a wiki page for developers regarding what the context menu category options and ordering for commands are. I think this would help our developers keep the context menus organized when adding new commands, as opposed to just adding them all to the bottom of the menu (which is what tends to happen currently).
>
> @zdmullen What do you think about this last point? Where should I set it up?

I know this question wasn't directed at me, but when I wrote my comment, I was thinking of putting the information in either [our repo's wiki](https://github.com/zowe/vscode-extension-for-zowe/wiki), or in [our repo's docs folder](https://github.com/zowe/vscode-extension-for-zowe/tree/master/docs).

zdmullen commented:
Wiki in the repo makes sense to me. Sorry for the delay

lauren-li commented:
@katelynienaber Now that the ability to favorite a profile/search has been re-added to the Data Sets view, I found that I can no longer remove a favorited profile/search. :P Can you also please re-add the context menu item for removing a favorited search in Data Sets? (Screenshot is from v1.11.1 release.) Thanks!:

<img width="317" alt="Screen Shot 2021-01-21 at 1 17 45 AM" src="https://user-images.githubusercontent.com/45975633/105291309-98121800-5b86-11eb-8e45-14c22f015204.png">

~
Also, regarding the wiki page, I was thinking it might just be something simple like this (but maybe also include the command order numbers, e.g. the `0` after the `@` symbol in `1_create@0`): https://github.com/zowe/vscode-extension-for-zowe/pull/127#issuecomment-505936284

katelynienaber commented:
Aaand again XD @lauren-li I've added this back.

zdmullen commented:
Just a short start of a page here: https://github.com/zowe/vscode-extension-for-zowe/wiki/Working-with-Context-Menus-and-Icons-in-Zowe-Explorer

lauren-li commented:
I went about clicking on the commands involved in this PR, and I found a bit of odd behavior when trying to favorite a profile/search in the Jobs view. Clicking "Add to Favorites" doesn't seem to add it into the Job view's Favorites section. However, when I next click on the profile/search, it will act like a profile/search item in the Favorites section, and show the context menu for a favorited profile/search, even though it is not located in the Favorites section:

![ze-jobFav2](https://user-images.githubusercontent.com/45975633/105362120-7c2f6600-5bc8-11eb-87f3-93b97cf64260.gif)


I also think we may want to reorganize the commands in `package.json` for better maintainability, but that should be a separate PR. I will start a discussion about it in our repo.

katelynienaber commented:
@lauren-li I've fixed the dropdown issue, please check again ^^

lauren-li commented:
@katelynienaber Thank you for addressing the issue I found. I don't see it happening anymore.

While manually retesting, I noticed a couple other issues. (Hopefully these will be the last ones!):
1. USS favorited profile nodes have extra commands in their context menus. The only command we currently support for these nodes is "Remove profile from Favorites". Clicking the other context menu items either doesn't do anything or produces errors, so they should be removed:
    <img width="349" alt="Screen Shot 2021-01-24 at 3 58 39 PM" src="https://user-images.githubusercontent.com/45975633/105644054-84560280-5e61-11eb-916e-f0909ade6814.png">
2. USS favorited profile nodes have the "Remove Favorite" star icon next to them. Clicking on it doesn't do anything, as that "Remove" command was not designed for favorited profile nodes, so the icon should be removed. (Favorited profile nodes in the Data Sets and Jobs views also do not have the star icon, so removing it here would also improve consistency between views.):
    <img width="248" alt="Screen Shot 2021-01-24 at 4 17 51 PM" src="https://user-images.githubusercontent.com/45975633/105644165-f4fd1f00-5e61-11eb-817d-d5f1b6934abe.png">
3. For job items in the Jobs view, we should add an "Add to Favorites" item into the context menu to be consistent with the way you can favorite items in the Data Sets and USS views:
    <img width="290" alt="Screen Shot 2021-01-24 at 4 23 32 PM" src="https://user-images.githubusercontent.com/45975633/105644245-aa2fd700-5e62-11eb-8a79-f893a3451948.png">

I think this PR is getting really close to being good to go. We are just at the polishing stage now. Thanks Katelyn for all your hard work on this!

katelynienaber commented:
@lauren-li This is like whack-a-mole XD Ok please give it another shot

lauren-li reviewed (changes requested):
Thanks @katelynienaber!

I made some suggestions for the `when... viewItem =~` values for USS Copy Path, USS Rename, and USS Pull from Mainframe, as it seems those commands got removed from some context menus where they should have remained. 

I like the addition of the inline "Remove Search" icons for favorited searches. Although I didn't ask for them since we never had them before, I did always think it was a bit odd we excluded them. The new icons seem to function well. I made some suggestions for the USS "Remove Search" icon to make it more consistent with MVS and Jobs.

I went ahead and made PR #1153 into your PR with these requested changes since I know your plate is already full with like 3 other pull requests that are a bit more complicated. Feel free to bring #1153 into this PR if you agree with the suggestions. I hope this helps, and that we can get this PR merged into `master` soon!

katelynienaber commented:
> Thanks @katelynienaber!
> 
> I made some suggestions for the `when... viewItem =~` values for USS Copy Path, USS Rename, and USS Pull from Mainframe, as it seems those commands got removed from some context menus where they should have remained.
> 
> I like the addition of the inline "Remove Search" icons for favorited searches. Although I didn't ask for them since we never had them before, I did always think it was a bit odd we excluded them. The new icons seem to function well. I made some suggestions for the USS "Remove Search" icon to make it more consistent with MVS and Jobs.
> 
> I went ahead and made PR #1153 into your PR with these requested changes since I know your plate is already full with like 3 other pull requests that are a bit more complicated. Feel free to bring #1153 into this PR if you agree with the suggestions. I hope this helps, and that we can get this PR merged into `master` soon!

@lauren-li Thanks for reviewing again!

I added these changes to this branch, but I would prefer not to merge your branch in. I noticed some weird behavior on #1153, ex. clicking a favorited search in USS adds an extra profile node, and the Remove Favorites icons are missing on the inline bar.

Maybe I was going too fast and built it wrong somehow. But these things are working on this branch so I added your changes here and would prefer to just merge this one (#957) without #1153.

Let me know what you think!

lauren-li commented:
> clicking a favorited search in USS adds an extra profile node

@katelynienaber I believe this is an existing issue (#1029 USS Favorited search creates extra profile node).

> Remove Favorites icons are missing on the inline bar

Oh no, sorry. That is somehow missing from my PR. Thanks for catching that.

Since you brought in the changes from PR #1153, I will go ahead and close it. I'll retest this PR (#957) and let you know how it goes!

lauren-li commented:
@katelynienaber I think this PR is almost there! I promise I'm not finding these issues one-at-a-time on purpose. All the issues I mentioned above are fixed, but I just noticed for sequential data sets only, the "Copy" and "Paste" commands are missing (for both favorited and non-favorited sequential data sets). The expected commands show fine for PDS's and members:

<img width="457" alt="Screen Shot 2021-01-25 at 11 17 35 AM" src="https://user-images.githubusercontent.com/45975633/105732912-f892b380-5efe-11eb-9a4f-c956f5526633.png">

Other than that, I did not notice any other issues. I counted the commands and compared them to the expected numbers for each item type this time, so I really hope this is the last bit of polish! All unit and expected integration tests are passing.

lauren-li commented:
> but I just noticed for sequential data sets only, the "Copy" and "Paste" commands are missing

@katelynienaber Actually, maybe we want to leave off the "Paste" command for sequential data sets. The current (v1.11.1) behavior is, if you "Copy" a data set member or other sequential data set, and then select "Paste" for a second sequential data set, the content of the second sequential data set gets completely replaced by the content of the copied data set. The mainframer I spoke with said this behavior does not seem intuitive, and I feel this is an easy way for users to accidentally lose data in sequential data sets.

So perhaps, let's re-add the "Copy" command for sequential data sets, but leave off the "Paste" command, if @jelaplan is ok with that?

katelynienaber commented:
> > but I just noticed for sequential data sets only, the "Copy" and "Paste" commands are missing
> 
> @katelynienaber Actually, maybe we want to leave off the "Paste" command for sequential data sets. The current (v1.11.1) behavior is, if you "Copy" a data set member or other sequential data set, and then select "Paste" for a second sequential data set, the content of the second sequential data set gets completely replaced by the content of the copied data set. The mainframer I spoke with said this behavior does not seem intuitive, and I feel this is an easy way for users to accidentally lose data in sequential data sets.
> 
> So perhaps, let's leave off the "Paste" command for sequential data sets (as you currently have in this PR), but re-add the "Copy" command for sequential data sets, if @jelaplan is ok with that?

We can discuss it in standup. The function exists in master and this PR is only about rearranging the context menus, not deciding whether any functions are useful or intuitive to users. But if everyone agrees that it is a useless function then I can remove it after standup so that we don't have to open another PR.

zdmullen commented:
Re-add Copy for sequential data sets (but leave out Paste, which you have taken out).

jellypuno reviewed (dismissed):
everything looks good! thank you @katelynienaber

lauren-li commented:
This looks really good! One last (I hope!) thing: Favorited profile nodes should only have "Remove Search" in their context menus. For the Jobs view only (not MVS or USS), there is an extra "Disable Validation for Profile" command:

<img width="336" alt="Screen Shot 2021-01-26 at 1 55 44 PM" src="https://user-images.githubusercontent.com/45975633/105891447-a7f28780-5fde-11eb-949a-cb4b8a43a03d.png">

katelynienaber commented:
@lauren-li Just pushed, check again pls ^^

lauren-li reviewed (approved):
@katelynienaber Thank you for all your hard work, patience, and perseverance on this PR. I think this looks good now!
